system/lib/wasmfs/paths.cpp

```diff
@@ -99,7 +99,20 @@ ParsedPath getParsedPath(std::vector<std::string> pathParts,
 
     bool atFinalPart = pathPart == (pathParts.end() - 1);
     if (atFinalPart) {
-      return ParsedPath{std::move(lockedCurr), child};
+      // We found the child to return, now compute the parent. Normally that is
+      // curr, but we must also handle the other cases of . and .. as well as
+      // the possible corner case of the child being moved or unlinked
+      // meanwhile.
```

Comment:
The child of `curr` cannot possibly have been moved or unlinked while we hold the lock on `curr`.

Reply:
If the current segment is `.` or `..` then we are not currently holding a lock on the parent.

Reply:
Added more comments, and a fix for the deadlock risk.